### PR TITLE
[ESQL] Add TO_DATE_NANOS conversion function

### LINKED ISSUE #111842: [ESQL] create TO_DATE_NANOS function
Similar to `TO_DATETIME`, but returning a date nanos type.  It should accept the same set of types as `TO_DATETIME`:

- [x]  `datetime`
- [x] `date_nanos`
- [x] `long`
- [x] `keyword`
- [x] `text`
- [x] `double`
- [x] `unsigned_long`
- ~[ ] `integer`~

### PR DESCRIPTION
**Pull request by not-napoleon**
Resolves https://github.com/elastic/elasticsearch/issues/111842

This adds a conversion function that yields `DATE_NANOS`.  Mostly this is straight forward.  

It is worth noting that when converting a millisecond date into a nanosecond date, the conversion function truncates it to 0 nanoseconds (i.e. first nanosecond of that millisecond).  This is, of course, a bit of an assumption, but I don't have a better assumption we can make.  I'd thought about adding a second, optional, parameter to control this behavior, but it's important that `TO_DATE_NANOS` extend `AbstractConvertFunction`, which itself extends `UnaryScalarFunction`, so that it will work correctly with union types.  Also, it's unlikely the user will have any better guess than we do for filling in the nanoseconds.

Making that assumption does, however, create some weirdness.  Consider two comparisons:

`TO_DATETIME("2023-03-23T12:15:03.360103847") == TO_DATETIME("2023-03-23T12:15:03.360")` will return true while `TO_DATE_NANOS("2023-03-23T12:15:03.360103847") == TO_DATE_NANOS("2023-03-23T12:15:03.360")` will return false.  This is akin to casting between longs and doubles, where things may compare equal in one type that are not equal in the other.  This seems fine, and I can't think of a better way to do it, but it's worth being aware of.

### CODE REVIEW

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/date_nanos.csv-spec [hunk 1]
```diff
@@ -70,3 +70,80 @@ FROM date_nanos | SORT millis asc | EVAL nanos = MV_LAST(nanos) | KEEP nanos | L
...
+ROW d = TO_DATE_NANOS("2023-03-23T12:15:03.360") | KEEP d;
+
+d:date_nanos
+2023-03-23T12:15:03.360Z
+;
+
+string to date nanos, out of range
+required_capability: date_nanos_type
+
+ROW d = TO_DATE_NANOS("2262-04-12T00:00:00.000") | KEEP d;
```
Comment (astefan) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/date_nanos.csv-spec:95]: I see the syntax `::date_nanos` is not supported (yet?). Is this intentional?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDateNanos.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        }
+        return nanos;
+    }
+
+    @ConvertEvaluator(extraName = "FromDatetime", warnExceptions = { IllegalArgumentException.class })
+    static long fromDatetime(long in) {
+        try {
+            return Math.multiplyExact(in, 1_000_000L);
+        } catch (ArithmeticException e) {
+            // This seems like a more useful error message than "Long Overflow"
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDateNanos.java:135]: 👍

**Review by bpintea — COMMENTED** [review]
We should add the type to the `EsqlDataTypeConverter.TYPE_TO_CONVERTER_FUNCTION` mapping and add a few tests with the `::` conversion operator.

**Review by astefan — COMMENTED** [review]
Added few comments. Nothing major. Also:
- some more tests in csv-spec would be welcomed:
  - a foldable function like `concat("2023-01-01","T12:12:12")` should be used as part of `eval` or `row` and its value computed in a `to_date_nanos` or `::date_nanos`.
  - tests with `null` as a constant and as a computed value: `to_date_nanos(null + 1::long)`, `eval x = null, y = concat("2023", null) | limit 1 | stats max(123) by to_date_nanos(y)`
  - some constant that fails the range check: `eval g = to_date_nanos(-5::long)`
  - tests with real indices data
- some tests with union types (`date` and `date_nanos` for example, `long` and `date_nanos`)
- `date_nanos` being a date data type, I think it's worth taking our date functions for a spin. I looked only at `date_extract("nano_of_day", to_date_nanos(z))` which fails with `second argument of [date_extract("nano_of_day", to_date_nanos(z))] must be [datetime], found value [to_date_nanos(z)] type [date_nanos]` and it shouldn't. Indeed, I didn't want to extend the support you are adding with this PR and have it more contained, I only tried to test the new function in nested functions calls

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDateNanos.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        if (in < 0d) {
+            throw new IllegalArgumentException("Nanosecond dates before 1970-01-01T00:00:00.000Z are not supported.");
+        }
+        return DataTypeConverter.safeDoubleToLong(in);
+    }
+
+    @ConvertEvaluator(extraName = "FromString", warnExceptions = { IllegalArgumentException.class })
+    static long fromKeyword(BytesRef in) {
+        Instant parsed = DateFormatters.from(DEFAULT_DATE_NANOS_FORMATTER.parse(in.utf8ToString())).toInstant();
+        long nanos = parsed.getEpochSecond();
```
Comment (ivancea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDateNanos.java:120]:
```suggestion
        long seconds = parsed.getEpochSecond();
```

**Comment by not-napoleon** [comment]
@astefan 

> some tests with union types (`date` and `date_nanos` for example, `long` and `date_nanos`)

I have a separate ticket for this, as I expect it will be quite involved. https://github.com/elastic/elasticsearch/issues/112885.  That said, if you'd rather I did it in this PR, I would be fine with that.  I just thought this PR was already getting pretty big.

> `date_nanos` being a date data type, I think it's worth taking our date functions for a spin. I looked only at `date_extract("nano_of_day", to_date_nanos(z))` which fails with `second argument of [date_extract("nano_of_day", to_date_nanos(z))] must be [datetime], found value [to_date_nanos(z)] type [date_nanos]` and it shouldn't. Indeed, I didn't want to extend the support you are adding with this PR and have it more contained, I only tried to test the new function in nested functions calls

Indeed, passing date nanos into date functions is not (yet) supported.  The roadmap for the feature lists that work further down: https://github.com/elastic/elasticsearch/issues/109352. At the moment we have it marked as not required for an MVP (i.e. before taking off the feature flag), but we can move it into the per-release work if we want to.  It will definitely be required if we implement the auto casting into date nanos that we talked about recently.

**Comment by astefan** [comment]
> I have a separate ticket for this, as I expect it will be quite involved. https://github.com/elastic/elasticsearch/issues/112885.

That's ok 👍

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDateNanos.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    @ConvertEvaluator(extraName = "FromDouble", warnExceptions = { IllegalArgumentException.class, InvalidArgumentException.class })
+    static long fromDouble(double in) {
+        if (in < 0d) {
+            throw new IllegalArgumentException("Nanosecond dates before 1970-01-01T00:00:00.000Z are not supported.");
+        }
+        return DataTypeConverter.safeDoubleToLong(in);
+    }
+
+    @ConvertEvaluator(extraName = "FromString", warnExceptions = { IllegalArgumentException.class })
+    static long fromKeyword(BytesRef in) {
```
[the remark this answers is not in the document]
Comment (not-napoleon) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDateNanos.java:122]: So, this conversion actually exists as `org.elasticsearch.common.time.DateUtils#toLong`, and in retrospect I should probably just delegate to that.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/date_nanos.csv-spec [hunk 5]
```diff
@@ -70,3 +70,80 @@ FROM date_nanos | SORT millis asc | EVAL nanos = MV_LAST(nanos) | KEEP nanos | L
...
+datetime to date nanos, with overflow
+required_capability: date_nanos_type
+
+ROW d = TO_DATE_NANOS(TO_DATETIME("2262-04-12T00:00:00.000")) | KEEP d; 
+warning:Line 1:9: evaluation of [TO_DATE_NANOS(TO_DATETIME(\"2262-04-12T00:00:00.000\"))] failed, treating result as null. Only first 20 failures recorded.
+warning:Line 1:9: java.lang.IllegalArgumentException: cannot create nanosecond dates after 2262-04-11T23:47:16.854775807Z
+
+d:date_nanos
+null
+;
```
[the remark this answers is not in the document]
Comment (not-napoleon) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/date_nanos.csv-spec:190]: I've added tests for converting field values.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/date_nanos.csv-spec
[the remark this answers is not in the document]
Comment (not-napoleon) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/date_nanos.csv-spec:1]: I've added row tests for pre-1970 dates.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java [hunk 6]
```diff
@@ -347,6 +348,7 @@ private FunctionDefinition[][] functions() {
                 def(ToCartesianShape.class, ToCartesianShape::new, "to_cartesianshape"),
                 def(ToDatePeriod.class, ToDatePeriod::new, "to_dateperiod"),
                 def(ToDatetime.class, ToDatetime::new, "to_datetime", "to_dt"),
+                def(ToDateNanos.class, ToDateNanos::new, "to_date_nanos", "to_datenanos"),
```
[the remark this answers is not in the document]
Comment (not-napoleon) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java:353]: Personally, I don't think so.  If I was going to add another alias, I'd opt for `date_nanoseconds`, to align with the title of https://www.elastic.co/guide/en/elasticsearch/reference/master/date_nanos.html, but really I think date_nanos is fine.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDateNanos.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        Instant parsed = DateFormatters.from(DEFAULT_DATE_NANOS_FORMATTER.parse(in.utf8ToString())).toInstant();
```
[the remark this answers is not in the document]
Comment (not-napoleon) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDateNanos.java:126]: I've updated this to use `DateUtils.toLong()` which does that check.

**Review by astefan — APPROVED** [review]
LGTM

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Pull request merged**